crates/sui-indexer-alt-framework/src/handlers/cp_sequence_numbers.rs

```diff
@@ -22,6 +23,80 @@ pub struct StoredCpSequenceNumbers {
 
 pub struct CpSequenceNumbers;
 
+/// A struct that can be instantiated by the pruner task to map a `from` and `to` checkpoint to its
+/// corresponding `tx_lo` and containing epoch. The `from` checkpoint is expected to be inclusive,
+/// and the `to` checkpoint is exclusive. This requires the existence of the `checkpoint_metadata`
+/// table.
+pub struct PrunableRange {
+    from: StoredCpSequenceNumbers,
+    to: StoredCpSequenceNumbers,
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum RangeError {
+    #[error("Invalid checkpoint range: `from` {0} is greater than `to` {1}")]
+    InvalidCheckpointRange(u64, u64),
+    #[error("No checkpoint mapping found for checkpoints {0} and {1}")]
+    NoCheckpointMapping(u64, u64),
+    #[error("Database error: {0}")]
+    DbError(diesel::result::Error),
+}
+
+impl PrunableRange {
+    /// Gets the tx and epoch mappings for both the start and end checkpoints.
+    ///
+    /// The values are expected to exist since the cp_mapping table must have enough information to
+    /// encompass the retention of other tables.
+    pub async fn get_range(
+        conn: &mut Connection<'_>,
+        from_cp: u64,
+        to_cp: u64,
+    ) -> Result<Self, RangeError> {
+        if from_cp > to_cp {
+            return Err(RangeError::InvalidCheckpointRange(from_cp, to_cp));
+        }
+
+        let results = cp_sequence_numbers::table
+            .select(StoredCpSequenceNumbers::as_select())
+            .filter(cp_sequence_numbers::cp_sequence_number.eq_any([from_cp as i64, to_cp as i64]))
+            .order(cp_sequence_numbers::cp_sequence_number.asc())
+            .load::<StoredCpSequenceNumbers>(conn)
+            .await
+            .map_err(RangeError::DbError)?;
+
+        let [first, last] = results.as_slice() else {
+            warn!("No checkpoint mapping found for checkpoint {from_cp} and {to_cp}. Found {} mapping(s) instead of 2", results.len());
+            return Err(RangeError::NoCheckpointMapping(from_cp, to_cp));
```

Comment:
`from_cp == to_cp` is a valid case where this slice pattern would not match.

Reply:
I think the simplest way to structure this code is to query using `eq_any` and then scan the results to find a match for each of `from_cp` and `to_cp`, something like the following, in the case of `tx_interval`:

```rust
let results: Vec<(i64, i64)> = cp_sequence_numbers::table
    .select((cp_sequence_numbers::cp_sequence_number, cp_sequence_numbers::tx_lo))
    .filter(cp_sequence_numbers::cp_sequence_number.eq_any([from_cp as i64, to_cp as i64])
    .load(conn)
    .await?;

let Some(from_tx) = results.iter().find_map(|(cp, tx)| (cp == from_cp).then(tx)) else {
    bail!("Could not find tx sequence number for checkpoint {from_cp}");
};

let Some(to_tx) = results.iter().find_map(|(cp, tx)| (cp == to_cp).then(tx)) else {
    bail!("Could not find tx sequence number for checkpoint {to_cp}"); 
};

Ok((from_tx..to_tx))
```

This also means you're left with only one error case, which further justifies simplifying the error story!

Reply:
Btw, in what scenario would `from_cp == to_cp` be a valid case?

`watermark.next_chunk` returns `to` as `(self.pruner_hi + size).min(self.reader_lo)`. `pruner_hi` and `reader_lo` should not be the same value, and `size` wouldn't make sense as a non-positive integer

Reply:
i suppose `from == to` is possible when we update `pruner_hi` on the pruning side before the committing side gets to update `reader_lo`

In that case though, it seems to be confusing to allow the pruning code to proceed since we expect `[from_cp, to_cp)`